Geo/line piping output

WPK4FEM opened the pull request:
**📝 Description**
Added
- output items PIPE_ACTIVE, PIPE_HEIGHT, FLUID_FLUX_VECTOR, LOCAL_FLUID_FLUX_VECTOR, PERMEABILITY_MATRIX, LOCAL_PERMEABILITY_MATRIX for the line piping element
- unit tests for those output items.

WPK4FEM committed:
CalculateOnIntegrationPoints first attempt + unit test for PIPE_ACTIVE

WPK4FEM committed:
CalculateOnIntegrationPoints first attempt + unit test for PIPE_ACTIVE

WPK4FEM committed:
Unit test for PIPE_HEIGHT added

WPK4FEM committed:
Unit test for PIPE_HEIGHT added

WPK4FEM committed:
Unit test for permeability constitutive tensor.

WPK4FEM committed:
Unit test for permeability constitutive tensor.

WPK4FEM committed:
std::fill_n and unit test for calculateOnIntegrationpoints of flux

WPK4FEM committed:
std::fill_n and unit test for calculateOnIntegrationpoints of flux

WPK4FEM committed:
SonarCloud + Clang Tidy remarks, Global flux output.

WPK4FEM committed:
SonarCloud + Clang Tidy remarks, Global flux output.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -32,12 +33,14 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
     explicit GeoSteadyStatePwPipingElement(IndexType NewId = 0) : Element(NewId) {}
 
-    GeoSteadyStatePwPipingElement(IndexType NewId, GeometryType::Pointer pGeometry)
+    GeoSteadyStatePwPipingElement(IndexType NewId, const GeometryType::Pointer& pGeometry)
```
I'm not sure how to comply with the Kratos Style Guide here. Perhaps:
```suggestion
    GeoSteadyStatePwPipingElement(IndexType NewId, const GeometryType::Pointer& rpGeometry)
```
???

WPK4FEM replied:
took the suggestion.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -32,12 +33,14 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
-    GeoSteadyStatePwPipingElement(IndexType NewId, GeometryType::Pointer pGeometry, PropertiesType::Pointer pProperties)
+    GeoSteadyStatePwPipingElement(IndexType                      NewId,
+                                  const GeometryType::Pointer&   pGeometry,
+                                  const PropertiesType::Pointer& pProperties)
```
Same thing here:
```suggestion
                                  const GeometryType::Pointer&   rpGeometry,
                                  const PropertiesType::Pointer& rpProperties)
```
???

WPK4FEM replied:
took the suggested change.

WPK4FEM committed:
std::size_t to avoid build warning that are treated as errors.

WPK4FEM committed:
Merge branch 'geo/line_piping_output_2D' of https://github.com/KratosMultiphysics/Kratos into geo/line_piping_output_2D

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+                this->GetGeometry().IntegrationPointsNumber(this->GetIntegrationMethod());
+            rValues.resize(number_of_integration_points);
+            auto        dynamic_viscosity_inverse = 1.0 / this->GetProperties()[DYNAMIC_VISCOSITY];
+            auto        constitutive_matrix = FillPermeabilityMatrix(this->GetValue(PIPE_HEIGHT));
```
These two local variables can be made `const`:
```suggestion
            const auto  dynamic_viscosity_inverse = 1.0 / this->GetProperties()[DYNAMIC_VISCOSITY];
            const auto  constitutive_matrix = FillPermeabilityMatrix(this->GetValue(PIPE_HEIGHT));
```

WPK4FEM replied:
done

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+                this->GetGeometry().IntegrationPointsNumber(this->GetIntegrationMethod());
+            rValues.resize(number_of_integration_points);
+            auto        dynamic_viscosity_inverse = 1.0 / this->GetProperties()[DYNAMIC_VISCOSITY];
+            auto        constitutive_matrix = FillPermeabilityMatrix(this->GetValue(PIPE_HEIGHT));
```
Is it better to name it as `permeability_matrix`?

WPK4FEM replied:
done.

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+                this->GetGeometry().IntegrationPoints(this->GetIntegrationMethod());
+            for (std::size_t i = 0; i < number_of_integration_points; ++i) {
+                // this should be rotated to the direction of the element for the global fluid flux vector
+                const auto head_gradient = CalculateHeadGradient(this->GetProperties(), this->GetGeometry());
```
It looks like `head_gradient` is calculated for the element. If this is correct, then it can be calculated together with `longitudinal_flux` before the for loop.

WPK4FEM replied:
In the current situation and for the 2 noded element you are very correct here. I did it because on a 3 node ( maybe to be created in fulure ) element it would be varying over the element.

Changed to once per element.

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+                    rValues[i][0] = longitudinal_flux * tangential_vector[0];
+                    rValues[i][1] = longitudinal_flux * tangential_vector[1];
+                    rValues[i][2] = longitudinal_flux * tangential_vector[2];
+                }
```
Would it be clearer to have the for loop (or std::fill) inside if statements? If `longitudinal_flux` is the same for all points, then  values are a tangential vector multiplied by the flux. In case of LOCAL_FLUID_FUX_VECTOR, the tangentional_vector is (1,0,0)

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -241,7 +318,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
-    Matrix FillPermeabilityMatrix(double pipe_height) const
+    static Matrix FillPermeabilityMatrix(double pipe_height)
     {
         return ScalarMatrix{1, 1, std::pow(pipe_height, 3) / 12.0};
```
Just for my knowledge, it should be power of 4 multiplied by a pipe factor, does it mean the factor is dimensional now? Thank you.

WPK4FEM replied:
Apparently yes, Jon and myself share your questioning.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+                                      std::vector<Matrix>&    rValues,
+                                      const ProcessInfo&      rCurrentProcessInfo) override
+    {
+        if (rVariable == PERMEABILITY_MATRIX || rVariable == LOCAL_PERMEABILITY_MATRIX) {
```
Given the implementation, the local and the global permeability matrices are identical...? I don't really understand why, sorry.

WPK4FEM replied:
1 x 1 matrix. You can rotate it to any angle, but it stays the same. This is the intrinsic permeability.

avdg81 replied:
Ah, yes, of course... Feeling a bit embarrassed. Thanks for explaining.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+        p_element->GetGeometry().IntegrationPointsNumber(p_element->GetIntegrationMethod()));
+    p_element->CalculateOnIntegrationPoints(PIPE_ACTIVE, pipe_active, dummy_process_info);
+    // Assert
+    KRATOS_EXPECT_FALSE(pipe_active[0])
```
I guess this implies we only have one integration point per piping element...? Perhaps we can assert that first...?
```cpp
    KRATOS_EXPECT_EQ(pipe_active_states.size(), 1);
```

WPK4FEM replied:
There are 2 integration points, which both should give the same.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+        CreateHorizontalUnitLengthGeoSteadyStatePwPipingElementWithPWDofs(r_model_part, p_properties);
+    p_element->GetProperties().SetValue(DENSITY_WATER, 1.0E3);
+    p_element->GetProperties().SetValue(DYNAMIC_VISCOSITY, 1.0E-2);
+    p_element->GetProperties().SetValue(PIPE_HEIGHT, 1.0E-1);
```
Sorry I forgot the difference between GetProperties().SetValue(PIPE_HEIGHT) and SetValue(PIPE_HEIGHT). Should this line be removed?

WPK4FEM replied:
The first writes to the material properties, the last directly to the element. Only the last is used, so I removed the first. Good catch!

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    p_element->SetValue(PIPE_ACTIVE, false);
+
+    std::vector<bool> pipe_active(
+        p_element->GetGeometry().IntegrationPointsNumber(p_element->GetIntegrationMethod()));
```
I would supply an empty vector the `CalculateOnIntegrationPoints` and then check whether the size is correct after calling it.
```suggestion
    std::vector<bool> pipe_active_states;
```

WPK4FEM replied:
Indeed, the CalculateOnIntegrationsPoints resizes the result itself.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    p_element->CalculateOnIntegrationPoints(LOCAL_FLUID_FLUX_VECTOR, fluid_fluxes, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_DOUBLE_EQ(fluid_fluxes[0](0), 1.E-3 * 0.1 / 12.0);
```
Should 1.E-3 and 0.1 have named constants?

markelov208 reviewed (commented):
Hi @WPK4FEM , I like how you made this PR because it is very understandable. Just a few comments.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    auto  p_element =
+        CreateHorizontalUnitLengthGeoSteadyStatePwPipingElementWithPWDofs(r_model_part, p_properties);
+
+    const double pipe_height = 1.234E-5;
```
```suggestion
    constexpr auto pipe_height = 1.234E-5;
```

WPK4FEM replied:
done

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    const double pipe_height = 1.234E-5;
+    p_element->SetValue(PIPE_HEIGHT, pipe_height);
+
+    std::vector<double> pipe_heights(
```
Also here, I would start with an empty vector. Let `CalculateOnIntegrationPoints` do the resizing when needed.

(I would like this suggestion to be propagated to the other tests, too, please.)

WPK4FEM replied:
done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    p_element->CalculateOnIntegrationPoints(PIPE_HEIGHT, pipe_heights, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_DOUBLE_EQ(pipe_heights[0], pipe_height);
```
And then also assert the size of `pipe_heights`.

(I would like this suggestion to be propagated to the other tests, too, please.)

WPK4FEM replied:
done

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    auto  p_element =
+        CreateHorizontalUnitLengthGeoSteadyStatePwPipingElementWithPWDofs(r_model_part, p_properties);
+
+    const double pipe_height = 1.E-1;
```
```suggestion
    constexpr auto pipe_height = 1.E-1;
```

WPK4FEM replied:
done

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    // Assert
+    KRATOS_EXPECT_DOUBLE_EQ(permeability_matrices[0](0, 0), std::pow(pipe_height, 3) / 12.0);
+
+    p_element->CalculateOnIntegrationPoints(LOCAL_PERMEABILITY_MATRIX, permeability_matrices, dummy_process_info);
```
Perhaps we should `clear` vector `permeability_matrices` first, to make sure that calling `CalculateOnIntegrationPoints` actually does something...?
```cpp
    permeability_matrices.clear();
```

WPK4FEM replied:
done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    p_element->GetGeometry()[0].FastGetSolutionStepValue(VOLUME_ACCELERATION) =
+        array_1d<double, 3>{0.0, -10.0, 0.0};
+    p_element->GetGeometry()[1].FastGetSolutionStepValue(VOLUME_ACCELERATION) =
+        array_1d<double, 3>{0.0, -10.0, 0.0};
```
Perhaps giving the vector a name avoids duplication and improves readability...?
```suggestion
    const auto gravity_acceleration = array_1d<double, 3>{0.0, -10.0, 0.0};
    p_element->GetGeometry()[0].FastGetSolutionStepValue(VOLUME_ACCELERATION) =
        gravity_acceleration;
    p_element->GetGeometry()[1].FastGetSolutionStepValue(VOLUME_ACCELERATION) =
        gravity_acceleration;
```

WPK4FEM replied:
done ( also on the earlier occasions. )

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_geo_steady_state_piping_element.cpp:
```diff
@@ -381,4 +381,124 @@ KRATOS_TEST_CASE_IN_SUITE(GeoSteadyStatePwPipingElementReturnsEquilibriumHeightF
+    KRATOS_EXPECT_DOUBLE_EQ(fluid_fluxes[0](0), 1.E-3 * 0.1 / 12.0);
+
+    // element tangetial axis is opposite global X, so global flux has negative X component
+    p_element->CalculateOnIntegrationPoints(FLUID_FLUX_VECTOR, fluid_fluxes, dummy_process_info);
```
Also here, I would `clear` vector `fluid_fluxes` prior to calling `CalculateOnIntegrationPoints` for the second time.

WPK4FEM replied:
done.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h:
```diff
@@ -140,6 +143,80 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoSteadyStatePwPipingElement : publ
+        }
+    }
+
+    void CalculateOnIntegrationPoints(const Variable<array_1d<double, 3>>& rVariable,
```
I have some suggestions on how we might simplify the body of this function. Perhaps we can discuss that when you are available?

WPK4FEM replied:
certainly.

avdg81 reviewed (commented):
A clear and very useful extension of the new piping element. I'm sure some people will be very pleased to have your changes available. All of my suggestions are minor in nature and non-blocking.

WPK4FEM committed:
First of review comments from Gennady taken up.

WPK4FEM committed:
Most of review suggestions by Anne.

avdg81 committed:
Refactored the calculation of the fluid flux vector

avdg81 committed:
Minor fixes and extensions

- Restored a variable name that was accidentally reverted.
- Check all entries of returned vectors.
- Fixed a typo.

avdg81 reviewed (approved):
Thank you for processing the review comments and replying to the questions. I believe this PR is ready to be merged.
